### DISCUSSION TITLE
Modify the available choices of the ListParameter while the application is running.

### DISCUSSION THREAD

**Discussion by Did-Mas** [pymeasure/pymeasure#1068] [Q&A] [answered]
I have ListParameters representing GPIB addresses of devices. I would like to be able to refresh the available options, but I don't see any way to change the choices field after defining the parameter.

**Comment by CasperSchippers** [comment, marked as answer]
I don't think there is a very straight-forward way to do this at the moment. When do you want to update the available options?

If it is just whenever you start the software, it is quite doable (bit hacky, but doable). You can update the `_choices` attribute (a dict containing string-representation <-> value pairs of the available options, but in your case will identical key-value paris since they are strings already) of the `ListParameter` in question BEFORE the `__init__` method of he ManagedWindow is called. using the `gui.py` as an example.
```python
class MainWindow(ManagedWindow):
    def __init__(self):

    # Can be be updated here
    TestProcedure.address._choices = {"GPIB::1": "GPIB::1", "GPIB::3": "GPIB::3"}

        super().__init__(
            procedure_class=TestProcedure,
            inputs=['iterations', 'delay', 'seed', 'address'],
            displays=['iterations', 'delay', 'seed'],
            x_axis='Iteration',
            y_axis='Random Number'
        )
        self.setWindowTitle('GUI Example')


if __name__ == "__main__":
    # Can be be updated here, I think over here would have my preference in a way
    TestProcedure.address._choices = {"GPIB::1": "GPIB::1", "GPIB::3": "GPIB::3"}

    app = QtWidgets.QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec())
```

If, on the other hand, you'd want to update it more often after the software is initialized..... that is way more difficult..... (and requires a helluvalot more 'hacking').

**Reply by CasperSchippers** [reply]
Actually, if you want to update it when you start the software, you could also define a function that collects the choices (e.g., `get_available_gpib_addresses`) and call that function where you define the `ListParameter` in you `Procedure`:
```python
address = ListParameter("GPIB address", choices=get_available_gpib_addresses())
```

**Reply by Did-Mas** [reply]
Thanks a lot for the answer. In the current version of my code, I set "choices", as per your second suggestion, during the parameter definition. However, I was wondering if it is possible to refresh the choices, for example, by pressing the "Refresh Addresses" button in case the device was not turned on when the program started.

**Reply by CasperSchippers** [reply]
I think the refresh button should definitely be possible, but it requires updating the choices in a number of places.

You could add a method like this to the MainWindow class, which should be connected to a button.
```python
def refresh_choices(self):
    # acquire the new addresses from the same function you used in the Parameter definition
    choices = get_available_gpib_addresses()

    # need to update the choices in the procedure, where only valid values can be set
    self.procedure_class.address._choices = dict(zip(choices, choices))

    # need to update the dropdown list
    self.inputs.address.clear()
    self.inputs.address.addItems(choices)
```

However, something is still missing.

**Reply by CasperSchippers** [reply]
I realized that I forgot that the address ListInput also keeps a copy of the parameter which needs to be updated. This seems to be working, just connect it to a button. 

```python
    def refresh_choices(self):
        # acquire the new addresses from the same function you used in the Parameter definition
        choices = get_available_gpib_addresses()  # Returns a list of options

        # need to update the choices in the procedure, where only valid values can be set
        self.procedure_class.address._choices = dict(zip(choices, choices))
        self.inputs.address._parameter._choices = dict(zip(choices, choices))

        # need to update the dropdown list
        self.inputs.address.clear()
        self.inputs.address.addItems(choices)
```

**Reply by Did-Mas** [reply]
Thanks a lot! I will try to implement it.
